Skip reading Parquet pages using Column Indexes feature of Parquet

JamesRTaylor opened the pull request:
Co-authored-by: Xinli Shang <shangx@uber.com>

Fixes https://github.com/trinodb/trino/issues/403

ryanrupp commented:
Can resolve https://github.com/trinodb/trino/issues/403

ryanrupp commented:
I haven't been following Trino Parquet changes recently but did the rewrite of the Parquet reader that @martint mention [here](https://github.com/trinodb/trino/issues/403#issuecomment-743420425) happen?

martint commented:
Not yet, I had to put it on the back burner due some other priorities, but I'll get back to it. We've decided we're going to try to get column index-based filtering in the current reader implementation.

martint commented on presto-parquet/src/main/java/io/prestosql/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -389,4 +502,353 @@ private static ParquetCorruptionException corruptionException(String column, Par
+                return Domain.create(ValueSet.all(type), hasNullValue);
+            }
+
+            if (i == 0) {
+                firstRange = range;
+            }
+            else {
+                restRanges[i - 1] = range;
+            }
```
This is a bit hacky. Let's add a `ValueSet.ofRanges(List<Range>)` method and just build a list in this method.

martint commented on presto-parquet/src/main/java/io/prestosql/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -389,4 +502,353 @@ private static ParquetCorruptionException corruptionException(String column, Par
+    private static <T extends Comparable<T>> Domain createDomain(Type type, ColumnIndex columnIndex, ParquetDataSourceId id, boolean hasNullValue, String columnName, List<T> mins, List<T> maxs)
+            throws ParquetCorruptionException
+    {
+        if (mins.size() == 0 || maxs.size() == 0 || mins.size() != maxs.size()) {
+            return Domain.create(ValueSet.all(type), hasNullValue);
+        }
```
When does this happen?

martint commented on presto-parquet/src/main/java/io/prestosql/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -389,4 +502,353 @@ private static ParquetCorruptionException corruptionException(String column, Par
+            if (min.compareTo(max) > 0) {
+                throw corruptionException(columnName, id, columnIndex);
+            }
+            //TODO:
```
remove

martint commented on presto-parquet/src/main/java/io/prestosql/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -389,4 +502,353 @@ private static ParquetCorruptionException corruptionException(String column, Par
+                throw corruptionException(columnName, id, columnIndex);
+            }
+            //TODO:
+            if (min instanceof Long) {
+                ParquetIntegerStatistics statistics = new ParquetIntegerStatistics((Long) min, (Long) max);
+                ranges.add((ParquetRangeStatistics<T>) statistics);
+            }
+            else if (min instanceof Double) {
+                ParquetDoubleStatistics statistics = new ParquetDoubleStatistics((Double) min, (Double) max);
+                ranges.add((ParquetRangeStatistics<T>) statistics);
+            }
+            else if (min instanceof Slice) {
```
This is a weird structure. The method is generic, but it switches on runtime type. The caller also switches on column type. It would be cleaner to have 3 methods (one for each type), or to abstract this logic via a function passed to this method.

jirassimok replied:
Especially because the types aren't primitive, I think the easiest solution is to have a parameter of type `BiFunction<? super T, ? super T, ? extends ParquetRangeStatistics<T>>`.

(But see my next comment.)

martint commented on presto-parquet/src/main/java/io/prestosql/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -389,4 +502,353 @@ private static ParquetCorruptionException corruptionException(String column, Par
+        return columnIndex.getMaxValues().size() == 0;
+    }
+
+    public FilterPredicate convertToParquetUdp()
```
What's a "Udp"? Don't abbreviate. Maybe call this `convertToParquetFilter`

martint commented on presto-parquet/src/main/java/io/prestosql/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -389,4 +502,353 @@ private static ParquetCorruptionException corruptionException(String column, Par
+    {
+        FilterPredicate filter = null;
+
+        // TODO: It could be a Presto bug that we don't see effectivePredicate.getDomains().get() has more than 1 domain.
+        //  For example, the 'where c1=3 or c1=10002' clause should have two domains but it has none
+        //  Todo: we assume the relation cross domains are 'or'
```
Is this a problem? If not, remove this TODO.

martint commented on presto-parquet/src/main/java/io/prestosql/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -389,4 +502,353 @@ private static ParquetCorruptionException corruptionException(String column, Par
+     */
+    static class DomainUserDefinedPredicate<T extends Comparable<T>>
+            extends UserDefinedPredicate<T>
+            implements Serializable
```
Why Serializable?

JamesRTaylor replied:
Being Serializable is required by argument of FilterApi.userDefined call below

martint commented on presto-parquet/src/main/java/io/prestosql/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -389,4 +502,353 @@ private static ParquetCorruptionException corruptionException(String column, Par
+            extends UserDefinedPredicate<T>
+            implements Serializable
+    {
+        private Domain columnDomain;
```
final

martint commented on presto-parquet/src/main/java/io/prestosql/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -389,4 +502,353 @@ private static ParquetCorruptionException corruptionException(String column, Par
+    {
+        private Domain columnDomain;
+
+        DomainUserDefinedPredicate(Domain domain)
```
`public`. It's part of this class' interface

martint commented on presto-parquet/src/main/java/io/prestosql/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -389,4 +502,353 @@ private static ParquetCorruptionException corruptionException(String column, Par
+
+    class ColumnIndexValueConverter
+    {
+        private final Map<String, Function<Object, Object>> ciConversions;
```
Don't abbreviate

martint commented on presto-parquet/src/main/java/io/prestosql/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -260,6 +310,59 @@ public static Domain getDomain(
+            return Domain.all(type);
+        }
+
+        if (descriptor.getType().equals(PrimitiveTypeName.INT32) || descriptor.getType().equals(PrimitiveTypeName.INT64) || descriptor.getType().equals(PrimitiveTypeName.FLOAT)) {
```
Also, there's some inconsistency (and potential discrepancy) between this, which switches on the parquet type, vs the code in ColumnIndexValueConverter, which switches on the Trino type. I think we may have the wrong abstraction for the conversions -- let me look more deeply.

JamesRTaylor replied:
For FLOAT, the code uses Float.floatToRawIntBits in ColumnIndexValueConverter.

martint replied:
That would be a problem, because the comparison as float values is not equivalent with the comparison of values after they are converted to their IEEE754 representation and interpreted as ints:

```
jshell> float a = -1E-10f, b = -1E20f; a < b; Float.floatToRawIntBits(a) < Float.floatToRawIntBits(b);
a ==> -1.0E-10
b ==> -1.0E20
$68 ==> false
$69 ==> true
```

Also, there needs to be special handling for NaN.

JamesRTaylor replied:
Definitely worthy of a test, but I believe Float.intBitsToFloat is used to convert back before comparisons are done.

martint replied:
That doesn't seem to be the case. `createDomain` just creates ranges using the raw values (as long).

JamesRTaylor replied:
Will add getMin/MaxValuesAsFloat . Or perhaps the getMin/MaxValuesAsDouble can be used in this case. Would be nice if the Type could be used instead of all these if statements. Do you think that's feasible?

martint replied:
Yes, possibly. See my other comment above about the potential inconsistencies between these checks and the ones in the converter, which are based on Type. I'll dig into this further as I continue reviewing, but it does seem like there's an opportunity for improving it.

JamesRTaylor replied:
This mechanism of converting a Float to int using `floatToRawIntBits` is already in use with statistics [here](https://github.com/trinodb/trino/blob/577302b514687331c9996fb8623b62dbf8c71b29/lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java#L207). Do you want me to change it there too?

JamesRTaylor replied:
Filed https://github.com/trinodb/trino/issues/8893 to improve this.

martint commented on presto-parquet/src/main/java/io/prestosql/parquet/reader/ColumnIndexStoreImpl.java:
```diff
@@ -0,0 +1,169 @@
+/**
+ * Internal implementation of {@link ColumnIndexStore}.
+ */
+public class ColumnIndexStoreImpl
```
Rename to TrinoColumnIndexStore or something like that. Don't use "Impl" in the name

martint reviewed (commented):
Some initial comments. I'm not done reviewing, yet.

JamesRTaylor commented on presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:
```diff
@@ -3314,6 +3315,7 @@ private void testMultipleWriters(Session session, HiveStorageFormat storageForma
                     createTableSql,
                     (long) computeActual("SELECT count(*) FROM tpch.sf1.orders").getOnlyValue());
 
+            computeActual("select custkey from scale_writers_large where custkey = 2");
```
I left this here as a way to reproduce the IllegalStateException thrown in LazyBlock (currently commented out and replaced with a println).

JamesRTaylor commented:
Thanks for the feedback so far, @martint. I'll definitely make all these changes. The main reason I posted this WIP patch was to:

1. Get feedback on overall approach.
2. Ask for insight/clues/ideas on why the IllegalStateException is occurring in LazyBlock so I can fix it (or get confirmation that it can be safely ignored with column indexing).
3. Get ideas on how to get adequate test coverage. Currently the test coverage is almost non existent. I tried reducing the Parquet page size so I could generate different scenarios with multiple pages, but it didn't seem to have an effect.

jirassimok commented on presto-parquet/src/main/java/io/prestosql/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -389,4 +502,353 @@ private static ParquetCorruptionException corruptionException(String column, Par
+            if (statistic == null) {
+                return false;
+            }
+            else {
+                if (statistic.getMin() instanceof Integer) {
```
As with `createDomain`, this `if`/`else` chain should be made more-generic. Unlike `createDomain`, we can't add a parameter to this method.

Instead, I might create a separate method like `<T> ParquetRangeStatistics<? super T> getRangeStatistics(T min, T max)`, which can be used both here and in `createDomain`.

jirassimok commented on presto-parquet/src/main/java/io/prestosql/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -389,4 +502,353 @@ private static ParquetCorruptionException corruptionException(String column, Par
+        @Override
+        public boolean inverseCanDrop(org.apache.parquet.filter2.predicate.Statistics<T> statistics)
+        {
+            // !canDrop() cannot be used because it might not be correct. To be safe, we just keep the record by returning false.
```
Why might `!canDrop` be incorrect?

Also, I'd split this comment onto two lines after the period.

jirassimok commented on presto-parquet/src/main/java/io/prestosql/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -389,4 +502,353 @@ private static ParquetCorruptionException corruptionException(String column, Par
+        {
+            //TODO: hasNullValue is set true. Would it generate false negative?
+            Domain domain = createDomain(columnDomain.getType(), true, parquetStatistics);
+            if (columnDomain.intersect(domain).isNone()) {
+                return true;
+            }
+            return false;
```
```suggestion
            return columnDomain.intersect(domain).isNone()
```

jirassimok commented on presto-parquet/src/main/java/io/prestosql/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -389,4 +502,353 @@ private static ParquetCorruptionException corruptionException(String column, Par
+        {
+            int pageCount = columnIndex.getMaxValues().size();
+            List<ByteBuffer> maxValues = columnIndex.getMaxValues();
+            List<Long> maxs = new ArrayList<>();
```
I think `maxes` is a better name.

I can't help but read `maxs` as `maxis` or `max s`.

This shows up in a few other places that I might have missed as well.

jirassimok commented on presto-parquet/src/main/java/io/prestosql/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -389,4 +502,353 @@ private static ParquetCorruptionException corruptionException(String column, Par
+            List<Long> mins = new ArrayList<>();
+            for (int i = 0; i < pageCount; i++) {
+                if (TINYINT.equals(type) || SMALLINT.equals(type) || INTEGER.equals(type)) {
+                    int minValue = converter.convert(minValues.get(i), column);
```
I'd probably inline the various `minValue`s and `maxValue`s in this file that are immediately added to a list and never used again.

jirassimok commented on presto-parquet/src/main/java/io/prestosql/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -389,4 +502,353 @@ private static ParquetCorruptionException corruptionException(String column, Par
+            return maxs;
+        }
+
+        public List<Double> getMinValuesAsDouble(Type type, ColumnIndex columnIndex, String column)
```
Most of these `get{Min,Max}ValuesAs[Type]` methods should be folded into a generic form to go with the suggested changes to `getDomain`.

I think you'll want something that looks mostly like this (but you may need to replace `getJavaType` to make it work).

```java
        public List<?> getMinValues(Type type, ColumnIndex columnIndex, String column)
        {
            return columnIndex.getMinValues().stream()
                    .map(value -> converter.convert(value, column))
                    .map(type.getJavaType()::cast)
                    .collect(Collectors.toList());
        }
```

jirassimok commented on presto-parquet/src/main/java/io/prestosql/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -389,4 +502,353 @@ private static ParquetCorruptionException corruptionException(String column, Par
+        }
+    }
+
+    class ColumnIndexValueConverter
```
This class should be private and static.

JamesRTaylor replied:
Marked as private, but the class can't be static because it uses the `converter` from the outer class.

jirassimok commented on presto-spi/src/main/java/io/prestosql/spi/block/LazyBlock.java:
```diff
@@ -380,7 +380,8 @@ private void load(boolean recursive)
             block = requireNonNull(loader.load(), "loader returned null");
             if (block.getPositionCount() != positionsCount) {
-                throw new IllegalStateException(format("Loaded block positions count (%s) doesn't match lazy block positions count (%s)", block.getPositionCount(), positionsCount));
+                System.out.println(format("Loaded block positions count (%s) doesn't match lazy block positions count (%s)", block.getPositionCount(), positionsCount));
```
Should either be `LOG.something` or removed.

jirassimok commented on presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergPageSourceProvider.java:
```diff
@@ -389,7 +389,7 @@ private static ConnectorPageSource createParquetPageSource(
                 long firstDataPage = block.getColumns().get(0).getFirstDataPageOffset();
                 if ((firstDataPage >= start) && (firstDataPage < (start + length)) &&
-                        predicateMatches(parquetPredicate, block, dataSource, descriptorsByPath, parquetTupleDomain)) {
+                        predicateMatches(parquetPredicate, block, dataSource, descriptorsByPath, parquetTupleDomain, null, false)) {
```
Instead of adding extra parameters here, it might also be a good idea to make an overload of `predicateMatches` without the new argument(s).

jirassimok reviewed (commented):
I've looked over most of this, and I've added comments on the hive and predicate packages (but not the reader package).

jirassimok commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:
```diff
@@ -332,6 +348,31 @@ public static ReaderPageSource createPageSource(
         return Optional.of(new GroupType(baseType.getRepetition(), baseType.getName(), ImmutableList.of(type)));
     }
 
+    private static ColumnIndexStore getColumnIndexStore(Predicate parquetPredicate, ParquetDataSource dataSource, BlockMetaData blockMetadata, Map<List<String>, RichColumnDescriptor> descriptorsByPath, boolean readColumnIndexFilter)
```
I'd name the third parameter just `block`, and wrap these parameters onto their own lines.

Once `readColumnIndexFilter` is part of `ParquetReaderOptions`, remove that parameter, make this method non-static, and use `this.options.getReadColumnIndexFilter()` instead.

jirassimok replied:
I believe this function should return `Optional<ColumnIndexStore>`. This goes along with my next comment.

jirassimok commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/TestHiveConfig.java:
```diff
@@ -101,7 +101,8 @@ public void testDefaults()
                 .setTimestampPrecision(HiveTimestampPrecision.DEFAULT_PRECISION)
                 .setOptimizeSymlinkListing(true)
-                .setLegacyHiveViewTranslation(false));
+                .setLegacyHiveViewTranslation(false)
+                .setReadColumnIndexFilter(true));
```
You get a smaller diff if you set this _above_ the existing last line, so the punctuation that's already there doesn't have to be moved. (This applies in a few other places as well.)

jirassimok commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/PredicateUtils.java:
```diff
@@ -122,14 +123,19 @@ public static Predicate buildPredicate(
     }
 
-    public static boolean predicateMatches(Predicate parquetPredicate, BlockMetaData block, ParquetDataSource dataSource, Map<List<String>, RichColumnDescriptor> descriptorsByPath, TupleDomain<ColumnDescriptor> parquetTupleDomain)
+    public static boolean predicateMatches(Predicate parquetPredicate, BlockMetaData block, ParquetDataSource dataSource, Map<List<String>, RichColumnDescriptor> descriptorsByPath, TupleDomain<ColumnDescriptor> parquetTupleDomain, ColumnIndexStore ciStore, boolean readColumnIndex)
```
The new arguments should be replaced with one argument of type `Optional<ColumnIndexStore>`. If it's empty, no index store should be used.

And the arguments can be put onto their own lines.

jirassimok commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:
```diff
@@ -332,6 +348,31 @@ public static ReaderPageSource createPageSource(
+            return null;
+        }
+
+        Set<ColumnPath> paths = new HashSet<>();
+        for (List<String> path : descriptorsByPath.keySet()) {
+            paths.add(ColumnPath.get(path.toArray(new String[0])));
+        }
+        return ColumnIndexStoreImpl.create(dataSource, blockMetadata, paths);
```
@martint's comment here:
> ```java
>         for (ColumnChunkMetaData column : blockMetadata.getColumns()) {
>             if (column.getColumnIndexReference() != null && column.getOffsetIndexReference() != null) {
>                 Set<ColumnPath> paths = descriptorsByPath.keySet().stream()
>                         .map(path -> ColumnPath.get(path.toArray(new String[0])))
>                         .collect(toImmutableSet());
>                 return ColumnIndexStoreImpl.create(dataSource, blockMetadata, paths);
>             }
>         }
>         return null;
> ```

jirassimok replied:
Or you could go all-in with streams, especially if the method returns an `Optional`:
```java
        return blockMetadata.getColumns().stream()
                .filter(column -> column.getColumnIndexReference() != null && column.getOffsetIndexReference() != null)
                .findAny()
                .map(ignored -> ColumnIndexStoreImpl.create(
                        dataSource,
                        blockMetadata,
                        descriptorsByPath.keySet().stream()
                                .map(path -> ColumnPath.get(path.toArray(String[]::new)))
                                .collect(toImmutableSet())));
```

jirassimok commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/PredicateUtils.java:
```diff
@@ -122,14 +123,19 @@ public static Predicate buildPredicate(
         }
 
+        // Page stats is finer grained but relatively more expensive, so we do the filtering after above block filtering.
+        if (ciStore != null && readColumnIndex && !parquetPredicate.matches(block.getRowCount(), ciStore, dataSource.getId())) {
+            return false;
+        }
```
With the `Optional` argument, this becomes
```java
if (indexStore.isPresent()
        && !parquetPredicate.matches(block.getRowCount(), indexStore.get(), dataSource.getId())) {
```
Unfortunately, the exception means a proper optional chain doesn't work.

jirassimok commented on lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetReader.java:
```diff
@@ -98,7 +122,10 @@ public ParquetReader(
             AggregatedMemoryContext systemMemoryContext,
-            ParquetReaderOptions options)
+            ParquetReaderOptions options,
+            Predicate parquetPredicate,
+            List<ColumnIndexStore> blockIndexStores,
```
To fit with the conventions of Trino, these nullable parameters should almost certainly be `Optional` instead.

jirassimok commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -127,6 +141,41 @@ public boolean matches(DictionaryDescriptor dictionary)
+            if (columnIndex == null || columnIndex.getMinValues().size() == 0 || columnIndex.getMaxValues().size() == 0 || columnIndex.getMinValues().size() != columnIndex.getMaxValues().size()) {
+                continue;
+            }
+            else {
```
The `else` isn't necessary.

jirassimok commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -368,11 +470,21 @@ private static ParquetCorruptionException corruptionException(String column, Par
 
+    private static ParquetCorruptionException corruptionException(String column, ParquetDataSourceId id, ColumnIndex columnIndex)
+    {
+        return new ParquetCorruptionException(format("Corrupted statistics for column \"%s\" in Parquet file \"%s\": [%s]", column, id, columnIndex));
```
I'd rephrase this as `Corrupted statistics for column "%s" in Parquet file "%s". Corrupted column index: %s` to make the last part as clear as possible.

jirassimok commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -389,4 +501,353 @@ private static ParquetCorruptionException corruptionException(String column, Par
+                range = Range.greaterThanOrEqual(type, function.apply(min));
+            }
+            else {
+                return Domain.create(ValueSet.all(type), hasNullValue);
```
This could use an explanatory comment. Why does it return early only in this case?

jirassimok commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -389,4 +501,353 @@ private static ParquetCorruptionException corruptionException(String column, Par
+
+    private boolean isCorruptedColumnIndex(ColumnIndex columnIndex)
+    {
+        if (columnIndex.getMaxValues() == null || columnIndex.getMinValues() == null ||
```
I'd put the || on the next line, so it's visible if you're looking down the left margin.

(Same a few lines down.)

jirassimok commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -389,4 +501,353 @@ private static ParquetCorruptionException corruptionException(String column, Par
+                throw corruptionException(columnName, id, columnIndex);
+            }
+            //TODO:
+            if (min instanceof Long) {
+                ParquetIntegerStatistics statistics = new ParquetIntegerStatistics((Long) min, (Long) max);
+                ranges.add((ParquetRangeStatistics<T>) statistics);
+            }
+            else if (min instanceof Double) {
+                ParquetDoubleStatistics statistics = new ParquetDoubleStatistics((Double) min, (Double) max);
+                ranges.add((ParquetRangeStatistics<T>) statistics);
+            }
+            else if (min instanceof Slice) {
+                ParquetStringStatistics statistics = new ParquetStringStatistics((Slice) min, (Slice) max);
+                ranges.add((ParquetRangeStatistics<T>) statistics);
+            }
```
@martint's comment here:
> This is a weird structure. The method is generic, but it switches on runtime type. The caller also switches on column type. It would be cleaner to have 3 methods (one for each type), or to abstract this logic via a function passed to this method.

jirassimok replied:
Especially because the types aren't primitive, I think the easiest solution is to have a parameter of type `BiFunction<? super T, ? super T, ? extends ParquetRangeStatistics<T>>`.

(But see my next comment.)

jirassimok commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -389,4 +501,353 @@ private static ParquetCorruptionException corruptionException(String column, Par
+            if (statistic == null) {
+                return false;
+            }
+            else {
+                if (statistic.getMin() instanceof Integer) {
```
As with `createDomain`, this `if`/`else` chain should be made more-generic. Unlike `createDomain`, we can't add a parameter to this method.

Instead, I might create a separate method like `<T> ParquetRangeStatistics<? super T> getRangeStatistics(T min, T max)`, which can be used both here and in `createDomain`.

Also, this particular `else` and `if` should be on the same line.

JamesRTaylor replied:
Fixed the `else`/`if`, but the creation of the ParquetRangeStatistics is different than in the previous case and the processing is slightly different in each case. I don't think there's much that can be done to clean it up more.

jirassimok replied:
I think the per-type processing could be split into a special class or method, similar to the use of `ParquetXStatistics` in `canDrop`, but to make it clean, it would probably need to be used more widely in the PR, which might not be worth the effort.

jirassimok commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -389,4 +501,353 @@ private static ParquetCorruptionException corruptionException(String column, Par
+        @Override
+        public boolean inverseCanDrop(org.apache.parquet.filter2.predicate.Statistics<T> statistics)
+        {
+            // !canDrop() cannot be used because it might not be correct. To be safe, we just keep the record by returning false.
```
Why might `!canDrop` be incorrect?

Also, I'd split this comment onto two lines after the period.

jirassimok commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -389,4 +501,353 @@ private static ParquetCorruptionException corruptionException(String column, Par
+        {
+            //TODO: hasNullValue is set true. Would it generate false negative?
+            Domain domain = createDomain(columnDomain.getType(), true, parquetStatistics);
+            if (columnDomain.intersect(domain).isNone()) {
+                return true;
+            }
+            return false;
```
```suggestion
            return columnDomain.intersect(domain).isNone()
```

jirassimok commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -389,4 +501,353 @@ private static ParquetCorruptionException corruptionException(String column, Par
+        {
+            int pageCount = columnIndex.getMaxValues().size();
+            List<ByteBuffer> maxValues = columnIndex.getMaxValues();
+            List<Long> maxs = new ArrayList<>();
```
I think `maxes` is a better name.

I can't help but read `maxs` as `maxis` or `max s`.

This shows up in a few other places that I might have missed as well.

jirassimok commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -389,4 +501,353 @@ private static ParquetCorruptionException corruptionException(String column, Par
+            List<Long> mins = new ArrayList<>();
+            for (int i = 0; i < pageCount; i++) {
+                if (TINYINT.equals(type) || SMALLINT.equals(type) || INTEGER.equals(type)) {
+                    int minValue = converter.convert(minValues.get(i), column);
```
I'd probably inline the various `minValue`s and `maxValue`s in this file that are immediately added to a list and never used again.

jirassimok commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -389,4 +501,353 @@ private static ParquetCorruptionException corruptionException(String column, Par
+            return maxs;
+        }
+
+        public List<Double> getMinValuesAsDouble(Type type, ColumnIndex columnIndex, String column)
```
Most of these `get{Min,Max}ValuesAs[Type]` methods should be folded into a generic form to go with the suggested changes to `getDomain`.

I think you'll want something that looks mostly like this (but you may need to replace `getJavaType` to make it work).

```java
        public List<?> getMinValues(Type type, ColumnIndex columnIndex, String column)
        {
            return columnIndex.getMinValues().stream()
                    .map(value -> converter.convert(value, column))
                    .map(type.getJavaType()::cast)
                    .collect(Collectors.toList());
        }
```

JamesRTaylor replied:
The processing is different across Long (switches on five different types), Double, and Slice.

jirassimok commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -389,4 +501,353 @@ private static ParquetCorruptionException corruptionException(String column, Par
+        }
+    }
+
+    class ColumnIndexValueConverter
```
This class should be private and static.

jirassimok commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:
```diff
@@ -472,7 +472,7 @@ private static ConnectorPageSource createParquetPageSource(
                 long firstDataPage = block.getColumns().get(0).getFirstDataPageOffset();
                 if ((firstDataPage >= start) && (firstDataPage < (start + length)) &&
-                        predicateMatches(parquetPredicate, block, dataSource, descriptorsByPath, parquetTupleDomain)) {
+                        predicateMatches(parquetPredicate, block, dataSource, descriptorsByPath, parquetTupleDomain, null, false)) {
```
Instead of adding extra parameters here, it might also be a good idea to make an overload of `predicateMatches` without the new argument(s).

jirassimok reviewed (commented):
(This is the same review I posted a short time ago, but on the rebased code. I've also copied a few of Martin's comments that I responded to.)

I've looked over most of this, and I've added comments on the hive and predicate packages (but not the reader package).

jirassimok commented on core/trino-spi/src/main/java/io/trino/spi/block/LazyBlock.java:
```diff
@@ -389,7 +389,8 @@ private void load(boolean recursive)
             block = requireNonNull(loader.load(), "loader returned null");
             if (block.getPositionCount() != positionsCount) {
-                throw new IllegalStateException(format("Loaded block positions count (%s) doesn't match lazy block positions count (%s)", block.getPositionCount(), positionsCount));
+                System.out.println(format("Loaded block positions count (%s) doesn't match lazy block positions count (%s)", block.getPositionCount(), positionsCount));
```
Should be `LOG.something` or removed.

JamesRTaylor commented on lib/trino-parquet/src/main/java/io/trino/parquet/reader/PrimitiveColumnReader.java:
```diff
@@ -166,10 +175,13 @@ public ColumnChunk readPrimitive(Field field)
             int valuesToRead = Math.min(remainingValueCountInPage, nextBatchSize - valueCount);
+            if (valuesToRead == 0) {
+                // When we break here, we could end up with valueCount < nextBatchSize, but that is OK.
+                break;
```
This break is what is causing the IllegalStateException to occur in LazyBlock, @martint. How should this be handled? Should we collect info about the values that are skipped inside of BlockBuilder and then take that into account in LazyBlock?

JamesRTaylor commented on presto-hive/src/main/java/io/prestosql/plugin/hive/parquet/HdfsParquetDataSource.java:
```diff
@@ -220,19 +227,41 @@ public void free()
 
-    private <K> Map<K, ChunkReader> readLargeDiskRanges(Map<K, DiskRange> diskRanges)
+    @Override
+    public ColumnIndex readColumnIndex(ColumnChunkMetaData column) throws IOException
```
I can move these methods, but I need to be able to call the Parquet library method Util.readColumnIndex(inputStream) to read the column index. Should I expose the underlying InputStream by adding a ParquetDataSource.getInputStream() method?

martint replied:
You can wrap the Slice returned by ChunkReader with an InputStream to pass to that method

JamesRTaylor replied:
I could change it to this instead:

    public ColumnIndex readColumnIndex(int offset, PrimitiveType type)

martint replied:
No, what I meant is that you can move these methods elsewhere so that this class has no knowledge about ColumnIndex or OffsetIndex. The caller would take the ChunkReader, get a Slice, wrap it with an InputStream and pass it to the `Util.readColumnIndex(...)` method.

JamesRTaylor replied:
See my comment here: https://github.com/trinodb/trino/pull/7349/files#r665764776

JamesRTaylor commented on presto-parquet/src/main/java/io/prestosql/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -127,6 +142,41 @@ public boolean matches(DictionaryDescriptor dictionary)
+            }
+
+            ColumnIndex columnIndex = ciStore.getColumnIndex(ColumnPath.get(column.getPath()));
+            if (columnIndex == null || columnIndex.getMinValues().size() == 0 || columnIndex.getMaxValues().size() == 0 || columnIndex.getMinValues().size() != columnIndex.getMaxValues().size()) {
```
`columnIndex` can be null if there's no information for the `columnPath`, for example if the column was added later or the column index store is corrupt. I've changed the check for size of 0 (which is a marker for the ColumnIndex being empty) to use the utility method.

JamesRTaylor commented on presto-parquet/src/main/java/io/prestosql/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -260,6 +310,59 @@ public static Domain getDomain(
     }
 
+    @VisibleForTesting
+    public Domain getDomain(Type type, long rowCount, ColumnIndex columnIndex, ParquetDataSourceId id, RichColumnDescriptor descriptor)
```
TestTupleDomainParquetPredicate is in a different package, so it needs to be public (like other getDomain declarations here)

martint replied:
Then we may want to move the tests for this class to the same package to avoid unnecessary exposure of these methods.

JamesRTaylor replied:
This class references JULIAN_EPOCH_OFFSET_DAYS which is package-protected in ParquetTimestampUtils. I could make that public and then move this class. Maybe in a separate PR?

JamesRTaylor commented on presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:
```diff
@@ -2462,7 +2463,7 @@ private void testInsert(Session session, HiveStorageFormat storageFormat)
         assertUpdate(session, "INSERT INTO test_insert_format_table (_double, _bigint) SELECT 2.72E0, 3", 1);
 
-        assertQuery(session, "SELECT * FROM test_insert_format_table WHERE _bigint = 3", "SELECT null, null, null, 3, null, null, null, null, 2.72, null, null, null");
+        assertQuery(session, "SELECT * FROM test_insert_format_table WHERE _double = CAST(2.72E0 as DOUBLE)", "SELECT null, null, null, 3, null, null, null, null, 2.72, null, null, null");
```
This causes the min/max Double code path to be executed in TupleDomainParquetPredicate, so it's related.

martint replied:
Are we losing some coverage for bigint by switching it, then?

jirassimok commented on lib/trino-parquet/src/main/java/io/trino/parquet/reader/ColumnIndexFilterUtils.java:
```diff
@@ -0,0 +1,171 @@
+    /*
+     * Returns the filtered offset index containing only the pages which are overlapping with rowRanges.
+     */
+    static FilteredOffsetIndex filterOffsetIndex(OffsetIndex offsetIndex, RowRanges rowRanges, long totalRowCount)
```
This could be a method of `OffsetIndex`, which could be a top-level class (making `ColumnIndexFilterUtils` unnecessary).

JamesRTaylor commented:
@martint - this is ready for review now. I've addressed the review comments and added a bunch of tests (and fixed a number of bugs that they uncovered). One minor limitation is that when column indexes on TIMESTAMP were used, they'd lead to that LazyBlock exception. I've worked around this by just not using column indexes in this case. If you're ok with this limitation, then this can be fixed in a separate PR.

JamesRTaylor commented:
@martint - the two failures don't look related to this PR. Please take a look when you have a chance.

martint commented on presto-hive/src/main/java/io/prestosql/plugin/hive/parquet/ParquetPageSourceFactory.java:
```diff
@@ -323,6 +339,31 @@ public static ReaderPageSourceWithProjections createPageSource(
+        for (List<String> path : descriptorsByPath.keySet()) {
+            paths.add(ColumnPath.get(path.toArray(new String[0])));
+        }
+        return ColumnIndexStoreImpl.create(dataSource, blockMetadata, paths);
```
That's an abuse of streams. The fact that it's ignoring the argument says there's something weird about that stream chain. It's using it just for conditional control flow, which is not the intended use of streams. Let's switch to the for loop variant, which makes the intent more obvious.

martint commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:
```diff
@@ -351,6 +363,28 @@ public static ReaderPageSource createPageSource(
+            Map<List<String>, RichColumnDescriptor> descriptorsByPath,
+            ParquetReaderOptions options)
+    {
+        if (!options.isUseColumnIndex() || !(parquetPredicate instanceof TupleDomainParquetPredicate)) {
```
This code shouldn't be aware of the type of `Predicate`, so remove the `!(parquetPredicate instanceof TupleDomainParquetPredicate)` part.

martint commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:
```diff
@@ -351,6 +363,28 @@ public static ReaderPageSource createPageSource(
+                                .map(path -> ColumnPath.get(path.toArray(String[]::new)))
+                                .collect(toImmutableSet())));
+    }
+
```
Change this to the form that uses the for loop. The fact the lambda ignores its argument is a hint that we're misusing stream chains for control flow conditional actions.

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -260,26 +306,88 @@ public static Domain getDomain(
 
-    private static Optional<ParquetIntegerStatistics> toParquetIntegerStatistics(Statistics<?> statistics, ParquetDataSourceId id, String column)
+    @VisibleForTesting
+    public Domain getDomain(Type type, ColumnIndex columnIndex, ParquetDataSourceId id, RichColumnDescriptor descriptor)
```
Move this method below so that the diff lines up better. Currently, it's trying to match `toParquetIntegerStatistics` with `getDomain` instead of `toParquetLongStatistics`.

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -260,26 +306,88 @@ public static Domain getDomain(
+        }
+        else if (statistics instanceof IntStatistics) {
             IntStatistics intStatistics = (IntStatistics) statistics;
-            if (intStatistics.genericGetMin() > intStatistics.genericGetMax()) {
```
This check seems to be gone. Why?

JamesRTaylor replied:
This is checked when the range is created deeper in the code.

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -260,26 +306,88 @@ public static Domain getDomain(
     {
-        if (statistics instanceof LongStatistics) {
-            LongStatistics longStatistics = (LongStatistics) statistics;
-            if (longStatistics.genericGetMin() > longStatistics.genericGetMax()) {
```
This check seems to be gone. Why?

JamesRTaylor replied:
This is checked when the range is created deeper in the code.

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/ParquetLongStatistics.java:
```diff
@@ -13,13 +13,13 @@
     private final Long maximum;
 
-    public ParquetIntegerStatistics(Long minimum, Long maximum)
+    public ParquetLongStatistics(Long minimum, Long maximum)
```
We may want to change the arguments to `long` in a separate commit. They are always non-null.

martint replied:
This doesn't seem to be addressed, yet.

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/PredicateUtils.java:
```diff
@@ -67,10 +68,10 @@
     private PredicateUtils() {}
 
-    public static boolean isStatisticsOverflow(Type type, ParquetIntegerStatistics parquetIntegerStatistics)
+    public static boolean isStatisticsOverflow(Type type, ParquetLongStatistics parquetLongStatistics)
```
I'd call the argument just `statistics`. No need for the overly verbose name.

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -175,72 +228,65 @@ public static Domain getDomain(
         if ((type.equals(BIGINT) || type.equals(TINYINT) || type.equals(SMALLINT) || type.equals(INTEGER)) && (statistics instanceof LongStatistics || statistics instanceof IntStatistics)) {
-            Optional<ParquetIntegerStatistics> parquetIntegerStatistics = toParquetIntegerStatistics(statistics, id, column);
-            if (parquetIntegerStatistics.isEmpty() || isStatisticsOverflow(type, parquetIntegerStatistics.get())) {
+            ParquetLongStatistics parquetIntegerStatistics = toParquetLongStatistics(statistics, id, columnName);
```
The variable name no longer matches the type. Also, call it just `longStatistics`. No need for the overly verbose name.

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -175,72 +228,65 @@ public static Domain getDomain(
-            return createDomain(type, hasNullValue, parquetIntegerStatistics.get(), statisticsValue -> {
-                if (((DecimalType) type).isShort()) {
-                    return statisticsValue;
+        if (type instanceof DecimalType) {
```
Why did the scale check get removed?

JamesRTaylor replied:
The scale check isn't required because if it's a short decimal, toParquetLongStatistics will convert it to a long and otherwise the else will handle it.

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -175,72 +228,65 @@ public static Domain getDomain(
+        if (type instanceof DecimalType) {
+            DecimalType decimalType = (DecimalType) type;
+            if (decimalType.isShort()) {
+                ParquetLongStatistics parquetLongStatistics = toParquetLongStatistics(statistics, id, columnName);
```
Call the variable `longStatistics`

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/PredicateUtils.java:
```diff
@@ -84,9 +85,9 @@ public static boolean isStatisticsOverflow(Type type, ParquetIntegerStatistics p
             DecimalType decimalType = (DecimalType) type;
-            if (!decimalType.isShort()) {
+            if (decimalType.getScale() > 0 || !decimalType.isShort()) {
                 // Smallest long decimal type with 0 scale has broader range than representable in long, as used in ParquetIntegerStatistics
```
The comment still refers to ParquetIntegerStatistics

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/ParquetTypeUtils.java:
```diff
@@ -238,4 +242,24 @@ public static long getShortDecimalValue(byte[] bytes)
+        return unscaledDecimal(value);
+    }
+
+    public static Slice getLongDecimalValue(ByteBuffer buffer)
```
This method isn't used

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/ParquetTypeUtils.java:
```diff
@@ -238,4 +242,24 @@ public static long getShortDecimalValue(byte[] bytes)
         return value;
     }
+
+    public static long getShortDecimalValue(ByteBuffer buffer)
```
This method isn't used

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -175,72 +228,65 @@ public static Domain getDomain(
-            }
-            ParquetStringStatistics parquetStringStatistics = new ParquetStringStatistics(minSlice, maxSlice);
-            return createDomain(type, hasNullValue, parquetStringStatistics);
+            ParquetSliceStatistics parquetSliceStatistics = new ParquetSliceStatistics(minSlice, maxSlice);
```
Call the variable `sliceStatistics`

martint replied:
This is still not addressed.

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -175,72 +228,65 @@ public static Domain getDomain(
-            }
-            ParquetIntegerStatistics parquetIntegerStatistics = new ParquetIntegerStatistics((long) intStatistics.getMin(), (long) intStatistics.getMax());
-            return createDomain(type, hasNullValue, parquetIntegerStatistics);
+            ParquetLongStatistics parquetLongStatistics = new ParquetLongStatistics((long) intStatistics.getMin(), (long) intStatistics.getMax());
```
`longStatistics`

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -260,26 +306,88 @@ public static Domain getDomain(
+            Slice maxSlice = Slices.wrappedBuffer(binaryStatistics.genericGetMax().getBytes());
+            long minValue = getShortDecimalValue(minSlice.getBytes());
+            long maxValue = getShortDecimalValue(maxSlice.getBytes());
+            return new ParquetLongStatistics(minValue, maxValue);
```
How do we know it's going to be a short decimal instead of a long decimal?

JamesRTaylor replied:
The caller is required to check the type.

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/reader/FilteredOffsetIndex.java:
```diff
@@ -0,0 +1,163 @@
+    /*
+     * Returns the filtered offset index containing only the pages which are overlapping with rowRanges.
+     */
+    static FilteredOffsetIndex filterOffsetIndex(OffsetIndex offsetIndex, RowRanges rowRanges, long totalRowCount)
```
`public`, since it's part of the public interface of this class.

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/reader/FilteredOffsetIndex.java:
```diff
@@ -0,0 +1,163 @@
+        }
+    }
+
+    List<OffsetRange> calculateOffsetRanges(long rowGroupOffset)
```
`public`, since it's part of the public interface of this class.

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/reader/FilteredOffsetIndex.java:
```diff
@@ -0,0 +1,163 @@
+    List<OffsetRange> calculateOffsetRanges(long rowGroupOffset)
+    {
+        List<OffsetRange> ranges = new ArrayList<>();
+        int n = this.getPageCount();
```
`n` -> `pageCount`

Also, `this.` is unnecessary.

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/reader/FilteredOffsetIndex.java:
```diff
@@ -0,0 +1,163 @@
+            for (int i = 0; i < n; ++i) {
+                long offset = this.getOffset(i);
+                int length = this.getCompressedPageSize(i);
+                // We need to adjust the offset by startingPosition for presto because dataSource.readFully() started at startingPosition
```
presto -> trino

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/reader/FilteredOffsetIndex.java:
```diff
@@ -0,0 +1,163 @@
+                long offset = this.getOffset(i);
+                int length = this.getCompressedPageSize(i);
+                // We need to adjust the offset by startingPosition for presto because dataSource.readFully() started at startingPosition
+                if (currentRange == null || !currentRange.extend(offset /*- startingPosition*/, length)) {
```
Remove commented out code

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/reader/FilteredOffsetIndex.java:
```diff
@@ -0,0 +1,163 @@
+    static class OffsetRange
+    {
+        private final long offset;
+        private long length;
```
It'd be safer and easier to read the code if this class were immutable, especially since values of OffsetRange are returned from the `calculateOffsetRanges` method.

This will require re-working the method above. Something like:

```
            ...
            long currentOffset = -1;
            long currentLength = 0;

            long firstPageOffset = offsetIndex.getOffset(0);
            // Add a range for the dictionary page if required
            if (rowGroupOffset < firstPageOffset) {
                // We need to adjust the offset by startingPosition for trino because dataSource.readFully() started at startingPosition
                currentOffset = rowGroupOffset;
                currentLength = firstPageOffset - rowGroupOffset;
            }

            for (int i = 0; i < pageCount; ++i) {
                long offset = getOffset(i);
                int length = getCompressedPageSize(i);

                if (currentOffset == -1) {
                    currentOffset = offset;
                    currentLength = length;
                }
                else if (currentOffset + currentLength == offset) {
                    currentLength += length;
                }
                else {
                    ranges.add(new OffsetRange(currentOffset, currentLength));
                    currentOffset = -1;
                    currentLength = 0;
                }
            }

            ranges.add(new OffsetRange(currentOffset, currentLength));
            
            ...
```

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/reader/FilteredOffsetIndex.java:
```diff
@@ -0,0 +1,163 @@
+        return ranges;
+    }
+
+    static class OffsetRange
```
`public`, since it's part of this class' public interface

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/reader/FilteredOffsetIndex.java:
```diff
@@ -0,0 +1,163 @@
+            }
+
+            for (int i = 0; i < n; ++i) {
+                long offset = this.getOffset(i);
```
`this.` is unnecessary

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/reader/FilteredOffsetIndex.java:
```diff
@@ -0,0 +1,163 @@
+        private final long offset;
+        private long length;
+
+        public OffsetRange(long offset, int length)
```
`long length`

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/reader/PageReader.java:
```diff
@@ -56,12 +64,14 @@ public DataPage readPage()
         }
         DataPage compressedPage = compressedPages.remove(0);
         try {
+            long firstRowIndex = getFirstRowIndex(pageIndex++, offsetIndex);
```
For clarity, move the `++` to a separate line. It's easier to reason about the code when there are no side-effects.

martint reviewed (commented):
Some more comments...

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetReader.java:
```diff
@@ -126,14 +163,40 @@ public ParquetReader(
+            ColumnDescriptor columnDescriptor = column.getColumnDescriptor();
+            this.paths.put(ColumnPath.get(columnDescriptor.getPath()), columnDescriptor);
+        }
+        if (options.isUseColumnIndex() && parquetPredicate instanceof TupleDomainParquetPredicate) {
```
The fact that we have to do the `instanceof` check and then downcast tells me there's something odd about the abstractions and structure of this code. Let me dig further to see how we can clean that up.

JamesRTaylor replied:
Let me know if you find a better way.

martint replied:
One option is to add a toParquetFilter() method to the Predicate interface, so this class doesn't need to know about the specific type of the Predicate instance. The method would return an `Optional<FilterPredicate>`

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetReader.java:
```diff
@@ -348,4 +455,25 @@ public AggregatedMemoryContext getSystemMemoryContext()
+            if (columnIndexStore.isPresent()) {
+                rowRanges = ColumnIndexFilter.calculateRowRanges(FilterCompat.get(filter), columnIndexStore.get(),
+                        paths.keySet(), blocks.get(blockIndex).getRowCount());
+                blockRowRanges.set(blockIndex, rowRanges);
```
Format with all arguments on the same line or each argument on a separate line.

JamesRTaylor commented on lib/trino-parquet/src/main/java/io/trino/parquet/DataPage.java:
```diff
@@ -17,11 +17,21 @@
+    }
+
+    /**
+     * @return the index of the first row index in this page or -1 if unset.
```
This value is the index of the first row in this page across all the rows for the column in the row group. It's read from the column index. It's used to help synchronize the reading across columns (in combination with the offset information in the column index). It's set to -1 in the PageReader.getFirstRowIndex() call if there's no offset index information (called by PageReader.readPage()).

Instead of having an unset value, it's probably cleaner to use an Optional<Long>.

JamesRTaylor commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/PredicateUtils.java:
```diff
@@ -84,9 +85,9 @@ public static boolean isStatisticsOverflow(Type type, ParquetIntegerStatistics p
         }
-        if (type instanceof DecimalType && ((DecimalType) type).getScale() == 0) {
+        if (type instanceof DecimalType) {
             DecimalType decimalType = (DecimalType) type;
-            if (!decimalType.isShort()) {
+            if (decimalType.getScale() > 0 || !decimalType.isShort()) {
```
This changed because this method can be called when a short decimal is used with a scale (for example the DECIMAL(12,2) test in TestParquetPageSkipping. Without this change, that test fails. It can be simplified further too by removing the  `decimalType.getScale() > 0` check above.

JamesRTaylor commented on lib/trino-parquet/src/main/java/io/trino/parquet/ParquetDataSource.java:
```diff
@@ -34,7 +37,11 @@
     Slice readFully(long position, int length);
 
-    <K> Map<K, ChunkReader> planRead(Map<K, DiskRange> diskRanges);
+    <K> Multimap<K, ChunkReader> planRead(Multimap<K, DiskRange> diskRanges);
```
Instead of reading one big chunk per column per row group, many smaller chunks are read (based on the column index information). The disk ranges are combined in sorted order (to determine if they are contiguous or near) and the multimap ensures values preserve insertion order.

martint replied:
Multimap doesn't guarantee order preservation. That's an artifact of the specific Multimap implementation. It would be clearer to use a `Map<K, List<...>>` to convey that.

JamesRTaylor replied:
Changed to ListMultimap which is documented to preserve insertion order of values.

JamesRTaylor commented on lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetColumnChunk.java:
```diff
@@ -75,22 +97,35 @@ public PageReader readAllPages()
+
+    private boolean hasMorePages(long valuesCountReadSoFar, int dataPageCountReadSoFar)
+    {
+        return offsetIndex == null ? valuesCountReadSoFar < descriptor.getColumnChunkMetaData().getValueCount()
```
We may be skipping a bunch of pages initially when column indexing is being used, so value count wouldn't account for that.

JamesRTaylor commented:
@martint - I pushed a new commit based on your review suggestions.

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/DataPage.java:
```diff
@@ -13,13 +13,15 @@
 {
     protected final int valueCount;
-    private final long firstRowIndex;
+    private final Optional<Long> firstRowIndex;
```
Use `OptionalLong`

martint commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:
```diff
@@ -364,25 +364,27 @@ public static ReaderPageSource createPageSource(
-                                .map(path -> ColumnPath.get(path.toArray(String[]::new)))
-                                .collect(toImmutableSet())));
+        for (ColumnChunkMetaData column : blockMetadata.getColumns()) {
+            if (column.getColumnIndexReference() != null && column.getOffsetIndexReference() != null) {
```
Isn't this check inverted? If any column chunk has a column index or offset index it will bail out.

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/reader/TrinoColumnIndexStore.java:
```diff
@@ -53,9 +54,12 @@
         {
             this.meta = meta;
-            OffsetIndex oi;
+            OffsetIndex oi = null;
```
Don't abbreviate variable names.

JamesRTaylor commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:
```diff
@@ -230,12 +237,15 @@ public static ReaderPageSource createPageSource(
+                Optional<ColumnIndexStore> columnIndex = getColumnIndexStore(dataSource, block, descriptorsByPath, options);
                 if (start <= firstDataPage && firstDataPage < start + length
-                        && predicateMatches(parquetPredicate, block, dataSource, descriptorsByPath, parquetTupleDomain)) {
+                        && predicateMatches(parquetPredicate, block, dataSource, descriptorsByPath, parquetTupleDomain, columnIndex)) {
```
The ColumnIndexStore is created before the ChunkReader here, so not sure how to implement your suggestion here: https://github.com/trinodb/trino/pull/7349#discussion_r665726893

JamesRTaylor commented:
FYI, I pushed a new commit with fixes for the review suggestions, @martint

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetReader.java:
```diff
@@ -295,6 +381,28 @@ private ColumnChunk readPrimitive(PrimitiveField field)
         return columnChunk;
     }
 
+    protected PageReader createPageReader(List<Slice> slices, ColumnChunkMetaData metadata, ColumnDescriptor columnDescriptor, OffsetIndex offsetIndex)
```
private

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetReader.java:
```diff
@@ -295,6 +381,28 @@ private ColumnChunk readPrimitive(PrimitiveField field)
+        return columnChunk.readAllPages();
+    }
+
+    protected List<Slice> allocateBlock(int fieldId)
```
private

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetReader.java:
```diff
@@ -348,4 +456,28 @@ public AggregatedMemoryContext getSystemMemoryContext()
+
+    private static <T> List<T> listWithNulls(int size)
+    {
+        return Stream.generate(() -> (T) null).limit(size).collect(Collectors.toCollection(ArrayList<T>::new));
```
Format like this for readability:
```suggestion
        return Stream.generate(() -> (T) null)
                .limit(size)
                .collect(Collectors.toCollection(ArrayList<T>::new));
```

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetReader.java:
```diff
@@ -348,4 +456,28 @@ public AggregatedMemoryContext getSystemMemoryContext()
+
+    private RowRanges getRowRanges(int blockIndex)
+    {
+        requireNonNull(filter, "filter is null");
```
If the method cannot be called when `filter` is null, it would be cleaner to model it as an argument to the method that can never be null. The caller would be responsible for calling the method only when it makes sense to do so and the intention would be clear.

As it stands, the method has hidden behaviors that depend on the current state of the object in a non-obvious way.

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/reader/PrimitiveColumnReader.java:
```diff
@@ -68,6 +73,9 @@
     private DataPage page;
     private int remainingValueCountInPage;
     private int readOffset;
+    private PeekingIterator<Long> indexIter;
```
Don't abbreviate variable names.

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/reader/PrimitiveColumnReader.java:
```diff
@@ -178,42 +192,127 @@ public ColumnChunk readPrimitive(Field field)
             }
             while (repetitionLevel != 0);
+
+            // TODO: We could created infinite loop
```
When would this happen? We should resolve that TODO

JamesRTaylor replied:
Will remove - it's no longer relevant.

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/reader/PrimitiveColumnReader.java:
```diff
@@ -178,42 +192,127 @@ public ColumnChunk readPrimitive(Field field)
+     * values (and the related rl and dl) for the rows [20, 39] in the end of the page 0 for col2. Similarly, we have to
+     * skip values while reading page0 and page1 for col3.
+     */
+    private void processValuesSync(int valuesToRead, Consumer<Void> valueConsumer)
```
Why not `Runnable` like the other method?

JamesRTaylor replied:
Fixed

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/reader/PrimitiveColumnReader.java:
```diff
@@ -309,10 +417,30 @@ private ValuesReader initDataReader(ParquetEncoding dataEncoding, int valueCount
         }
     }
+
+    private boolean nextRow(int rl)
```
Don't abbreviate variable names.

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/reader/PrimitiveColumnReader.java:
```diff
@@ -309,10 +417,30 @@ private ValuesReader initDataReader(ParquetEncoding dataEncoding, int valueCount
+            if (currentRow > targetRow) {
+                targetRow = indexIter.hasNext() ? indexIter.next() : Long.MAX_VALUE;
+            }
+            return currentRow++ < targetRow;
```
Separate the `++` from the comparison for clarity. It's harder to reason about side effects and ordering of operations the way it's written.

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/reader/PrimitiveColumnReader.java:
```diff
@@ -178,42 +192,127 @@ public ColumnChunk readPrimitive(Field field)
+        for (int i = 0; i < valuesToRead; ) {
+            boolean consumed = false;
+            do {
+                if (nextRow(repetitionLevel)) {
```
I'm not sure I understand the semantics of the `nextRow` method. What does it mean for it to return `true` vs `false`? Why do we skip a value when it returns `true`?

JamesRTaylor replied:
Added comment. We skip rows until the currentRow reaches the targetRow.

JamesRTaylor replied:
Also, renamed function to make it more clear what it's doing.

JamesRTaylor commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -174,116 +235,150 @@ public static Domain getDomain(
-        if ((type.equals(BIGINT) || type.equals(TINYINT) || type.equals(SMALLINT) || type.equals(INTEGER)) && (statistics instanceof LongStatistics || statistics instanceof IntStatistics)) {
-            Optional<ParquetIntegerStatistics> parquetIntegerStatistics = toParquetIntegerStatistics(statistics, id, column);
-            if (parquetIntegerStatistics.isEmpty() || isStatisticsOverflow(type, parquetIntegerStatistics.get())) {
+        if ((type.equals(BIGINT) || type.equals(TINYINT) || type.equals(SMALLINT) || type.equals(INTEGER) || type.equals(DATE)) && min instanceof Number) {
```
That check is mimicing the former statistics instanceof check. I can remove that if you think it's not necessary.

JamesRTaylor replied:
The only time I can think of in which that instanceof check is needed would be if stats were generated and then the schema was changed in some incompatible way. An alternative would be to remove the instanceof checks and catch ClassCastException in toParquetLongStatistics and throw corruptionException instead to indicate that the statistics are corrupt.

martint replied:
I think that would be preferable.

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -174,116 +233,150 @@ public static Domain getDomain(
+
+    private static ParquetLongStatistics toParquetLongStatistics(Object min, Object max)
+    {
+        if (min instanceof Number) {
+            Number minValue = (Number) min;
+            Number maxValue = (Number) max;
+            return new ParquetLongStatistics(minValue.longValue(), maxValue.longValue());
+        }
```
It'd be better to check for the possible concrete types explicitly. `longValue` on `Number` can cause loss of precision for some data types such as Double. If Double is not allowed or expected, the method shouldn't even try to handle it.

JamesRTaylor replied:
Added check for Double or Float and throw now instead

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/reader/PrimitiveColumnReader.java:
```diff
@@ -178,42 +191,126 @@ public ColumnChunk readPrimitive(Field field)
+     * page.
+     * For example:
+     *
+     * rows   col1   col2   col3
+     *      ┌──────┬──────┬──────┐
+     *   0  │  p0  │      │      │
+     *      ╞══════╡  p0  │  p0  │
+     *  20  │ p1(X)│------│------│
+     *      ╞══════╪══════╡      │
+     *  40  │ p2(X)│      │------│
+     *      ╞══════╡ p1(X)╞══════╡
+     *  60  │ p3(X)│      │------│
+     *      ╞══════╪══════╡      │
+     *  80  │  p4  │      │  p1  │
+     *      ╞══════╡  p2  │      │
+     * 100  │  p5  │      │      │
+     *      └──────┴──────┴──────┘
```
This should be wrapped in `<pre> ... </pre>` so that the javadoc renders correctly

JamesRTaylor replied:
Done

martint commented on lib/trino-parquet/src/main/java/io/trino/parquet/reader/PrimitiveColumnReader.java:
```diff
@@ -178,42 +191,126 @@ public ColumnChunk readPrimitive(Field field)
+     * values (and the related rl and dl) for the rows [20, 39] in the end of the page 0 for col2. Similarly, we have to
+     * skip values while reading page0 and page1 for col3.
+     */
+    private void processValuesSync(int valuesToRead, Runnable valueReader)
```
This method and `processValues` look very similar. What's the actual difference? Can the other one be implemented in terms of this one?

JamesRTaylor replied:
Good call. I've consolidated them.

JamesRTaylor commented:
Pushed another commit based on the recent feedback. Thanks so much for the review, @martint.

martint commented:
Merged, thanks!

findepi commented:
> Refactor code to compute Domain based on Statistics


just curious, what was the refactor's objective?

JamesRTaylor commented:
> just curious, what was the refactor's objective?

Code simplification. There were three separate long if/then/else blocks when there only needed to be a single one. Also, the number of instanceof checks were reduced as well.

losipiuk commented:
@JamesRTaylor @martint is there a reason to keep https://github.com/trinodb/trino/issues/403 open with this one merged?
